### ISSUE TITLE
barplot showing average reads per treatment

### ISSUE THREAD

**Issue by ademenez** [joey711/phyloseq#437]
Hi,
I am new to Phyloseq but lately I have played around a lot with some tree, ordination and bar plot features and have found it very useful. I wonder if, for display purposes, whether there is a way to plot average abundance of sequences per taxa per treatment, and perhaps with the inclusion of standard error bars? Currently I know I can merge abundances for several samples in one treatment, but so far I have only managed to plot the sum of reads and not averages, and I haven't found a way to show how variable the read numbers are per treatment.

Thanks for the help,

Alex

**Comment by audy** [comment]
First you need to transform your read counts into percentages:

```S
data(GlobalPatterns)
transform_sample_counts(GlobalPatterns, function(x) 100*x/sum(x))
```

The easiest way to create box-plots is to first convert your phyloseq object into a data frame using `psmelt` then just plot a box-plot using ggplot2.

```
# Reduce the number of taxa because plotting 19k is too much
ss <- subset_taxa(GlobalPatterns, Phylum %in% c('Bacteroidetes', 'Proteobacteria'))
glom <- tax_glom(ss, taxrank = 'Phylum')

dat <- psmelt(glom) # create data frame

# boxplots! (replace SampleType with treatment)
ggplot(dat, aes(x=Phylum, y=Abundance, fill=SampleType)) + geom_boxplot()
```

![screenshot 2015-03-06 12 34 39](https://cloud.githubusercontent.com/assets/115904/6530248/2b273c92-c3fd-11e4-9506-8533520a5f7c.png)

**Comment by ademenez** [comment]
Hi Austin,

Many thanks! I'll give it a go.

Alex

**Comment by ClaireA** [comment]
Hi,

I've tried to followed this example but I get the following error after I try to confer the phyloseq object into a data frame:

Error in order(mdf$Abundance, decreasing = TRUE) : 
  argument 1 is not a vector

Any idea of how to fix it?

Thanks,

Claire

**Comment by audy** [comment]
I can't tell without seeing `mdf$Abundance`. Can you reply with the output of `head(mdf$Abundance)`?

**Comment by audy** [comment]
Hi Claire,

I sent you a reply here
https://github.com/joey711/phyloseq/issues/437#issuecomment-77636329. Not
sure if you got an email notification about it.

--austin

On Thu, Mar 26, 2015 at 6:17 PM, ClaireA <notifications@github.com> wrote:

> Hi,
>
> I've tried to followed this example but I get the following error after I
> try to confer the phyloseq object into a data frame:
>
> Error in order(mdf$Abundance, decreasing = TRUE) :
> argument 1 is not a vector
>
> Any idea of how to fix it?
>
> Thanks,
>
> Claire
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/joey711/phyloseq/issues/437#issuecomment-86736988>.
>



-- 
--austin

**Comment by audy** [comment]
My hunch is that `tax_glom` didn't work because your taxonomy table doesn't have a `Phylum` rank.

**Comment by ClaireA** [comment]
Here is an example of what I am getting:

> head(clairef1vlrf$Abundance)
Error in clairef1vlrf$Abundance : 
  $ operator not defined for this S4 class
> rank_names(clairef1vlrf)
[1] "Domain"  "Phylum"  "Class"   "Order"   "Family"  "Genus"   "Species" "Strain" 

Thanks for the help!

**Comment by audy** [comment]
You need to "melt" the Phyloseq object into a data.frame using the `psmelt` function.

```S
glom <- tax_glom(clairef1vlrf, taxrank = 'Phylum')
mdat <- melt(glom)

head(mdat$Abundance)
```

**Comment by ClaireA** [comment]
It is working now. For whatever reason as soon as I restarted ny R session it started working again. One question left, what is the significance of a black circle above or below a bar in this type of box plot?
Many thanks,
Claire

**Comment by audy** [comment]
Those are outliers

--austin

**Comment by joey711** [comment]
This is now fixed in `1.13.3+`.
If you don't want to install the devel version (simply wait for the next BioC release), then @michberr 's `detach(reshape)` workaround should also be fine. This is due to a name collision with the old, slow "reshape" package. phyloseq uses the faster "reshape2" package of the same author. They both contain a function called `melt`. This issue is now fixed, as phyloseq explicitly calls `reshape2::melt` in all instances.

Will close for now. Can re-open if this issue resurfaces.

joey
